### ISSUE TITLE
Race in Gloo/Gateway and notification stream with virtualservice status propagation

### ISSUE THREAD

**Issue by kdorosh** [solo-io/gloo#2313]
The flaky test that highlights the suspected bugs: https://github.com/solo-io/gloo/blob/master/test/kube2e/gateway/gateway_test.go#L580

> It("adds the invalid virtual services back into the proxy when updating an upstream makes them valid"

The last line is the flaky one. We often see:
```gateway tests with virtual service with a mix of valid and invalid virtual services [It] adds the invalid virtual services back into the proxy when updating an upstream makes them valid
/Users/kdorosh/go/src/github.com/solo-io/gloo/test/kube2e/gateway/gateway_test.go:583

  Timed out after 10.003s.
  Expected
      <core.Status_State>: 1
  to equal
      <core.Status_State>: 2

  /Users/kdorosh/go/src/github.com/solo-io/gloo/test/kube2e/gateway/gateway_test.go:628

  Full Stack Trace
  github.com/solo-io/gloo/test/kube2e/gateway_test.glob..func1.2.6.5()
  	/Users/kdorosh/go/src/github.com/solo-io/gloo/test/kube2e/gateway/gateway_test.go:628 +0x16c3
  github.com/onsi/ginkgo/internal/leafnodes.(*runner).runSync(0xc000259500, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
  	/Users/kdorosh/go/pkg/mod/github.com/onsi/ginkgo@v1.11.0/internal/leafnodes/runner.go:113 +0xe0
  github.com/onsi/ginkgo/internal/leafnodes.(*runner).run(0xc000259500, 0x3, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
```

The test:
- creates petstore service & deployment
- creates an invalid virtual service that points to an upstream with a function
- ensures that virtual service gets marked as invalid
- enables FDS for the discovered petstore upstream
- FDS updates upstream spec with the function referenced by the virtual service
- ensures that virtual service status gets marked as accepted

The last step is the flaky one.

This looks like two bugs:
- if you comment out the hash-checking in the gateway api emitter (https://github.com/solo-io/gloo/blob/master/projects/gateway/pkg/api/v1/api_snapshot_emitter.sk.go#L284-L286) you will resync the gateway every second, which eliminates most instances of the flake (it appears we are dropping a notification from gloo to gateway with a force emit here somehow)
- even with a forced update or constant resyncs, sometimes we hit the flake, e.g.: https://github.com/solo-io/gloo/pull/2281

**Comment by kdorosh** [comment]
Further, with both gloo and gateway writing status reports we see races between them writing the status on the proxy/virtual service, as well as running up the generation of the proxy by cycling through so many updates.

**Comment by CamelCaseNotation** [comment]
Does this affect routes being available after virtualservice creation, or is this only about the gloo control plane notifying/updating a custom resource's status? Or something else?

**Comment by kdorosh** [comment]
@CamelCaseNotation  After further digging, we have identified the cause.

We had a leaking goroutine in `projects/gloo/pkg/validation/server.go` that kept open streams to older validators. This created a race where gateway could subscribe to changes on a stream that was no longer receiving updates in gloo.

Thus changes to Gloo-watched resources would never trigger a resync in Gateway if Gateway itself didn't watch those resources (e.g., changes to upstream, which Gloo watches but Gateway does not, make a virtual service valid). This could affect the propagation of virtual service configuration to the proxy if validation was enabled.

**Comment by CamelCaseNotation** [comment]
I think I can reproduce this, or it at least I think it's happening to me in a soak test I'm running with Gloo. 

This definitely affects the routes and when they become useable. Would like to know more about this issue and the progress on it

**Comment by kdorosh** [comment]
The fix for this issue was released in 1.2.17 and 1.3.4, in open source. Gloo Enterprise 1.2.4 is the first enterprise release with this fix.

@CamelCaseNotation can you try your soak test on one of these versions and report back?

**Comment by rickducott** [comment]
Closing as fixed -- let's reopen if we encounter any more issues.
